# Analyzing profitable apps on the App Store and Google Play Markets

In this project we are assuming we are a data analyst working for a company that builds Android and iOS Mobile apps. We only build free apps so the primary source of revenue is in-app ads. This means that we must track the numbr of users who use the app.

The goal of this project is to identify apps types of apps that are profitable for the company!

In [1]:
from csv import reader

In [2]:
# Apple Store Data
opened_file = open('AppleStore.csv')
read_file = reader(opened_file)
ios = list(read_file)
ios_header = ios[0]
ios = ios[1:]

# Google Play Data
opened_file = open('googleplaystore.csv')
read_file = reader(opened_file)
android = list(read_file)
android_header = android[0]
android = android[1:]

In [3]:
def explore_data(dataset, start, end, rows_and_columns=False):
    data_slice = dataset[start:end]
    for element in data_slice:
        print(element)
        print('\n')
        
    if rows_and_columns:
        print('Number of rows:', len(dataset))
        print('Number of columns:', len(dataset[0]))

In [4]:
print(android_header)
print('\n')
explore_data(android, 1, 3, True)



print(ios_header)
print('n')
explore_data(ios, 1, 3, True)


['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', 'January 15, 2018', '2.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


Number of rows: 10841
Number of columns: 13
['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']
n
['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.

# Google Play
Number of rows: 10841
Number of columns: 13
The interesting columns are: App, Ratings, Reviews, Price, and Installs


# App Store
Number of rows: 7197
Number of columns: 16
The interesting columns are: id, price, rating_count_tot, and  user_rating

In [5]:
print(android[10472])

['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up']


Missing the "category" column so we delete row 10472

In [6]:
del android[10472]

In [7]:
print(android[10472])

['osmino Wi-Fi: free WiFi', 'TOOLS', '4.2', '134203', '4.1M', '10,000,000+', 'Free', '0', 'Everyone', 'Tools', 'August 7, 2018', '6.06.14', '4.4 and up']


In [8]:
duplicatesAnd = []
uniquesAnd = []

for app in android:
    name = app[0]
    if name in uniquesAnd:
        duplicatesAnd.append(name)
    else:
        uniquesAnd.append(name)


print('Number of Android Duplicates: ', len(duplicatesAnd))
print('\n')
print('Examples of Duplicates:', duplicatesAnd[:10])


Number of Android Duplicates:  1181


Examples of Duplicates: ['Quick PDF Scanner + OCR FREE', 'Box', 'Google My Business', 'ZOOM Cloud Meetings', 'join.me - Simple Meetings', 'Box', 'Zenefits', 'Google Ads', 'Google My Business', 'Slack']


In [9]:
duplicatesIos = []
uniquesIos = []

for app in ios:
    track_name = app[0]
    if track_name in uniquesIos:
        duplicatesIos.append(track_name)
    else:
        uniquesIos.append(track_name)
        
print('Number of Ios Duplicates', len(duplicatesIos))
print('\n')
print('Examples of Ios Duplicates', duplicatesIos[:10])

Number of Ios Duplicates 0


Examples of Ios Duplicates []


We won't remove duplicates randomly we will look at which duplicate is the most recent by looking at the number of ratings to determine which one will stay.

In [10]:
print(android_header)
print(android[1])

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']
['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', 'January 15, 2018', '2.0.0', '4.0.3 and up']


In [11]:
reviews_max = {}

for apps in android:
    name = apps[0]
    n_reviews = float(apps[3])
    if name in reviews_max and (reviews_max[name] < n_reviews):
        reviews_max[name] = n_reviews
    if name not in reviews_max:
        reviews_max [name] = n_reviews


a = len(reviews_max)
print(a)

9659


In [12]:
android_clean = []
already_added = []

for elements in android:
    name = elements[0]
    n_reviews = float(elements[3])
    if (reviews_max[name] == n_reviews) and (name not in already_added):
        android_clean.append(elements)
        already_added.append(name)
        
explore_data(android_clean, 0, 3, True)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


Number of rows: 9659
Number of columns: 13


# We use English for apps that we develop so we only want to analyze apps that are in English.

English characters are in range 0-127 for the ASCII character system so we will set a contraint to filter out apps that are greater than 127.

In [13]:
def filter(string):
    non_ascii = 0
    for element in string:
        if ord(element) > 127:
            non_ascii += 1
    if non_ascii > 3:
        return False
    else:
        return True

In [14]:
filter('Docs To Go™ Free Office Suite')

True

In [15]:
filter('Instachat 😜')

True

In [16]:
English_android = []
English_ios = []

for element in android_clean:
    name = element[0]
    if filter(name):
        English_android.append(element)

for element in ios:
    name = element[1]
    if filter(name):
        English_ios.append(element)

explore_data(English_android,1,3,True)
print('\n')
explore_data(English_ios,1,3,True)

['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


Number of rows: 9614
Number of columns: 13


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Games', '38', '5', '18', '1']


Number of rows: 6183
Number of columns: 16


In [17]:
print(ios_header)

['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


In [18]:
print(android_header)

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


In [19]:
Free_AND = []
Free_IOD = []

for elements in English_android:
    price = elements[7]
    if price == '0':
        Free_AND.append(elements)

for elements in English_ios:
    price = elements[4]
    if price == '0.0':
        Free_IOD.append(elements)
        
print(len(Free_AND))
print('\n')
print(len(Free_IOD))

8864


3222


In [20]:
explore_data(Free_AND, 0, 5)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


['Pixel Draw - Number Art Coloring Book', 'ART_AND_DESIGN', '4.3', '967', '2.8M', '100,000+', 'Free', '0', 'Everyone', 'Art & Design;Creativity', 'June 20, 2018', '1.1', '4.4 and up']


['Paper flowers instructions', 'ART_AND_DESIGN', '4.4', '167', '5.6M', '50,000+', 'Free', '0', 'Everyone', 'Art & Design', 'March 26, 2017', '1.0', '2.3 and up']




In [21]:
explore_data(Free_IOD, 0, 5)

['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Games', '38', '5', '18', '1']


['420009108', 'Temple Run', '65921024', 'USD', '0.0', '1724546', '3842', '4.5', '4.0', '1.6.2', '9+', 'Games', '40', '5', '1', '1']


['284035177', 'Pandora - Music & Radio', '130242560', 'USD', '0.0', '1126879', '3594', '4.0', '4.5', '8.4.1', '12+', 'Music', '37', '4', '1', '1']




In [22]:
Common_Genre_AND = {}
Common_Genre_IOS = {}
total_AND = 0
total_IOS = 0

for element in Free_AND:
    total_AND += 1
    genre = element[1]
    if genre in Common_Genre_AND:
        Common_Genre_AND[genre] += 1
    else:
        Common_Genre_AND[genre] = 1

for element in Free_IOD:
    total_IOS += 1
    genre = element[11]
    if genre in Common_Genre_IOS:
        Common_Genre_IOS[genre] += 1
    else:
        Common_Genre_IOS[genre] = 1

print(Common_Genre_AND)
print('\n')
print(Common_Genre_IOS)
print('\n')
        
AND_percentages = {}
IOS_percentages = {}

for key in Common_Genre_AND:
    percentage = (Common_Genre_AND[key]/total_AND)*100
    AND_percentages[key] = percentage

for key in Common_Genre_IOS:
    percentage = (Common_Genre_IOS[key]/total_IOS)*100
    IOS_percentages[key] = percentage

print(AND_percentages)
print('\n')
print(IOS_percentages)



{'WEATHER': 71, 'GAME': 862, 'FOOD_AND_DRINK': 110, 'PERSONALIZATION': 294, 'SOCIAL': 236, 'FINANCE': 328, 'LIBRARIES_AND_DEMO': 83, 'TOOLS': 750, 'AUTO_AND_VEHICLES': 82, 'COMMUNICATION': 287, 'TRAVEL_AND_LOCAL': 207, 'HOUSE_AND_HOME': 73, 'FAMILY': 1676, 'SPORTS': 301, 'VIDEO_PLAYERS': 159, 'EDUCATION': 103, 'PHOTOGRAPHY': 261, 'COMICS': 55, 'ART_AND_DESIGN': 57, 'LIFESTYLE': 346, 'PARENTING': 58, 'PRODUCTIVITY': 345, 'MEDICAL': 313, 'ENTERTAINMENT': 85, 'SHOPPING': 199, 'DATING': 165, 'HEALTH_AND_FITNESS': 273, 'BEAUTY': 53, 'EVENTS': 63, 'BOOKS_AND_REFERENCE': 190, 'MAPS_AND_NAVIGATION': 124, 'NEWS_AND_MAGAZINES': 248, 'BUSINESS': 407}


{'Lifestyle': 51, 'Social Networking': 106, 'Business': 17, 'Games': 1874, 'Productivity': 56, 'Reference': 18, 'Shopping': 84, 'Food & Drink': 26, 'Entertainment': 254, 'Catalogs': 4, 'Education': 118, 'Navigation': 6, 'Finance': 36, 'News': 43, 'Book': 14, 'Utilities': 81, 'Medical': 6, 'Photo & Video': 160, 'Weather': 28, 'Music': 66, 'Health & 

In [23]:
def freq_table(dataset, index):
    table = {}
    total = 0
    
    for element in dataset:
        total += 1
        value = element[index]
        if value in table:
            table[value] += 1
        else:
            table[value] = 1
    
    table_percentages = {}
    
    for element in table:
        percentage = (table[element]/total)*100
        table_percentages[element] = percentage
    
    return table_percentages

In [24]:
def display_table(dataset, index):
    table = freq_table(dataset, index)
    table_disp = []
    for key in table:
        key_tuple = (table[key], key)
        table_disp.append(key_tuple)
    
    table_sorted = sorted(table_disp, reverse = True)
    for entry in table_sorted:
        print(entry[1], ':', entry[0])

In [25]:
print(android_header)
print('\n')
print(ios_header)

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


In [26]:
# Category Android
print(display_table(Free_AND, 1))
print('\n')
# Genres Android
print(display_table(Free_AND, 9))
print('\n')
# prime_genre IOS
print(display_table(Free_IOD, -5))

FAMILY : 18.907942238267147
GAME : 9.724729241877256
TOOLS : 8.461191335740072
BUSINESS : 4.591606498194946
LIFESTYLE : 3.9034296028880866
PRODUCTIVITY : 3.892148014440433
FINANCE : 3.7003610108303246
MEDICAL : 3.531137184115524
SPORTS : 3.395758122743682
PERSONALIZATION : 3.3167870036101084
COMMUNICATION : 3.2378158844765346
HEALTH_AND_FITNESS : 3.0798736462093865
PHOTOGRAPHY : 2.944494584837545
NEWS_AND_MAGAZINES : 2.7978339350180503
SOCIAL : 2.6624548736462095
TRAVEL_AND_LOCAL : 2.33528880866426
SHOPPING : 2.2450361010830324
BOOKS_AND_REFERENCE : 2.1435018050541514
DATING : 1.861462093862816
VIDEO_PLAYERS : 1.7937725631768955
MAPS_AND_NAVIGATION : 1.3989169675090252
FOOD_AND_DRINK : 1.2409747292418771
EDUCATION : 1.1620036101083033
ENTERTAINMENT : 0.9589350180505415
LIBRARIES_AND_DEMO : 0.9363718411552346
AUTO_AND_VEHICLES : 0.9250902527075812
HOUSE_AND_HOME : 0.8235559566787004
WEATHER : 0.8009927797833934
EVENTS : 0.7107400722021661
PARENTING : 0.6543321299638989
ART_AND_DESIGN : 

# What is the most common genre in the prime_genre column of the App Store data set?

Games accounts for 58% of the data set!

# What other patterns?

Entertainment and Social Networking account for approximately another 10%

# General Impressions

The most common genres are all related to entertainment

# We can't recommend an app profile based on this frequency table alone because this is only for free apps

# For the Google Play Data the most common categories are family then games then tools

# The most common Genres are tools and entertainment

# I would recommend games and entertainment because they were part of the most common free apps for the IOS and Android!


# Calculate the average number of installs for each app genre

In [27]:
freq_table(Free_IOD, -5)

{'Book': 0.4345127250155183,
 'Business': 0.5276225946617008,
 'Catalogs': 0.12414649286157665,
 'Education': 3.662321539416512,
 'Entertainment': 7.883302296710118,
 'Finance': 1.1173184357541899,
 'Food & Drink': 0.8069522036002483,
 'Games': 58.16263190564867,
 'Health & Fitness': 2.0173805090006205,
 'Lifestyle': 1.5828677839851024,
 'Medical': 0.186219739292365,
 'Music': 2.0484171322160147,
 'Navigation': 0.186219739292365,
 'News': 1.3345747982619491,
 'Photo & Video': 4.9658597144630665,
 'Productivity': 1.7380509000620732,
 'Reference': 0.5586592178770949,
 'Shopping': 2.60707635009311,
 'Social Networking': 3.2898820608317814,
 'Sports': 2.1415270018621975,
 'Travel': 1.2414649286157666,
 'Utilities': 2.5139664804469275,
 'Weather': 0.8690254500310366}

In [30]:
genres_ios = freq_table(Free_IOD, -5)

for genre in genres_ios:
    total = 0
    len_genre = 0
    for apps in Free_IOD:
        genre_app = apps[-5]
        if genre_app == genre:
            num_ratings = float(apps[5])
            total += num_ratings
            len_genre += 1
    avg_n_ratings = total/len_genre
    print(genre, ":" , avg_n_ratings)

Lifestyle : 16485.764705882353
Social Networking : 71548.34905660378
Business : 7491.117647058823
Games : 22788.6696905016
Productivity : 21028.410714285714
Reference : 74942.11111111111
Shopping : 26919.690476190477
Food & Drink : 33333.92307692308
Entertainment : 14029.830708661417
Catalogs : 4004.0
Weather : 52279.892857142855
Navigation : 86090.33333333333
Finance : 31467.944444444445
News : 21248.023255813954
Book : 39758.5
Utilities : 18684.456790123455
Medical : 612.0
Photo & Video : 28441.54375
Education : 7003.983050847458
Music : 57326.530303030304
Health & Fitness : 23298.015384615384
Travel : 28243.8
Sports : 23008.898550724636


I would recommend Social Networking because it has a high average rating and it is one of the most common genres


# Average number of installs per app genre for Google Play data set

In [37]:
category_GP = freq_table(Free_AND, 1)

for category in category_GP:
    total = 0
    len_category = 0
    for apps in Free_AND:
        category_app = apps[1]
        if category_app == category:
            num_installs = apps[5]
            num_installs = num_installs.replace('+','')
            num_installs = num_installs.replace(',', '')
            total += float(num_installs)
            len_category += 1
    avg_installs = total/len_category
    print(category, ":", avg_installs )

WEATHER : 5074486.197183099
GAME : 15588015.603248259
FOOD_AND_DRINK : 1924897.7363636363
BEAUTY : 513151.88679245283
SOCIAL : 23253652.127118643
FINANCE : 1387692.475609756
LIBRARIES_AND_DEMO : 638503.734939759
AUTO_AND_VEHICLES : 647317.8170731707
BOOKS_AND_REFERENCE : 8767811.894736841
ENTERTAINMENT : 11640705.88235294
HOUSE_AND_HOME : 1331540.5616438356
FAMILY : 3695641.8198090694
SPORTS : 3638640.1428571427
VIDEO_PLAYERS : 24727872.452830188
LIFESTYLE : 1437816.2687861272
SHOPPING : 7036877.311557789
COMICS : 817657.2727272727
ART_AND_DESIGN : 1986335.0877192982
TOOLS : 10801391.298666667
PARENTING : 542603.6206896552
PRODUCTIVITY : 16787331.344927534
MEDICAL : 120550.61980830671
TRAVEL_AND_LOCAL : 13984077.710144928
PHOTOGRAPHY : 17840110.40229885
DATING : 854028.8303030303
HEALTH_AND_FITNESS : 4188821.9853479853
PERSONALIZATION : 5201482.6122448975
EVENTS : 253542.22222222222
COMMUNICATION : 38456119.167247385
MAPS_AND_NAVIGATION : 4056941.7741935486
EDUCATION : 1833495.14563106